### PR TITLE
Adds ability to set timeouts for a knxdclient.KNXDConnection

### PR DESCRIPTION
**Pull request by mrctrifork**
First of all, thank you for creating this library. It's nice and easy to use.

## Issue
I have a server running `knxd` that reboots once a day and I've noticed that if one uses `iterate_group_telegrams` and `knxd` reboots it leaves the coroutine hanging; awaiting for a message that will never come since the connection was dropped. This is caused because there were no timeouts in place.

## Suggested fix
I've added an optional parameter to the class `KNXDConnection` that can be used as a second argument to `asyncio.wait_for` and it is used in places where we are reading from `knxd`. The timeout errors will allow us to break out of the `while True` polling statements.

If the `timeout` is `None` old behavior is used instead.

Thank you for taking a look at this.

Miquel

### CODE REVIEW

File: knxdclient/client.py
Comment (mhthies) [inline knxdclient/client.py]: Please add a return type annotation here. Then, we can remove the inline comment, I think.

File: knxdclient/client.py
Comment (mhthies) [inline knxdclient/client.py]: Hm. I don't think, we should apply the timeout again separately here. Instead, IMHO it would be better to notify this coroutine when the `run()` coroutine exits (for any reason), via an `asyncio.Event` or similar.

Same is true for the `open_group_socket()` method below. Both coroutines could wait for the same `self._run_task_exited` Event (or similar), when waiting for the normal result, using something like
```python
get_task = asyncio.create_task(queue.get())
wait_exit = asyncio.create_task(self._run_task_exited.wait())
done, _ = asyncio.wait((get_task, wait_exit), return_when=asyncio.FIRST_COMPLETED)
if get_task not in done:
    raise ...
```

File: knxdclient/client.py
Comment (mhthies) [inline knxdclient/client.py:160]: It would be cool to have some kind of "keep alive" feature here, i.e. trigger some request to KNXD every few seconds or so, so that we can be sure to receive a packet from KNXD regularly. Then we don't have to rely on KNX bus activity to avoid false timeouts and can chose an even shorter timeout value.

I will take a look at the KNXD protocol spec to see if there is a function that we can (mis)use as a keepalive.

**Review by mhthies — CHANGES_REQUESTED** [review]
First, Thank you very much for your contribution! :)

I like the general approach and I noticed that I have had the unaddressed TODO comment to add a timeout for the open_group_socket method all along.

I've left a few comments, because I'd like to have the architecture, how the timeouts are implemented, a bit different. After writing the comments, I'm now wondering whether we need this timeout at all: Shouldn't the TCP connection be closed, when your server reboots, such that we get an `asyncio.IncompleteReadError` and the `run()` coroutine raises an exception? If that doesn't work, maybe we should try to configure the TCP keepalive on the network socket? 

We'd still need some internal Event to let the other coroutines fail with some Exception as well (see my inline comment for more details), instead of waiting infinitely for a packet on a broken socket connection. But it feels cleaner to rely on the operating system's detection of connection loss than applying our own (duplicated) timeout logics.

**Comment by mrctrifork** [comment]
Hi, thanks a lot for your feedback.

I am newish writing python. My background is more TypeScript / Kotlin during these last years. I lack knowledge on how async/await actually works in python so I just attempted to make it work.


In regards this:

> After writing the comments, I'm now wondering whether we need this timeout at all: Shouldn't the TCP connection be closed, when your server reboots, such that we get an asyncio.IncompleteReadError and the run() coroutine raises an exception? If that doesn't work, maybe we should try to configure the TCP keepalive on the network socket?

Well – Your judgement will be better than mine. My programming background is mostly in other languages so I just don't know the python internals about async/await. I just figured that the "program got stuck because the queue was not closed yet the connection got dropped".

I'll go over your comments when I get the chance and ping you once the changes have been submitted.

Again; thank you for your time and feedback and I'll take this as far as I can :)

File: knxdclient/client.py
Comment (mhthies) [reply knxdclient/client.py:160]: Hm. Unfortunately, according to the BCUSDK docs about EIBD's/KNXD's interface, there is no appropriate function to use as a keepalive, that reliably sends a response packet, once we opened the 'group socket' connection.

I'm somehow unconvinced to rely on KNX bus traffic to avoid running into timeouts.

Have you tested whether the event notification mechanism, you implemented today, is sufficient to detect the connection loss and stop the receive loop, when your KNXD server reboots – without the timeout here in the `run()` coroutine (or with a very very long timeout)?
If it doesn't work: Can you test whether setting up TCP keepalive for the socket here allows us to detect the connection loss reliably?
```python
sock = self._writer.get_extra_info("socket")
# The following seems to depend on your OS and is specific for Linux. See https://stackoverflow.com/q/12248132/10315508
sock.setsockopt(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)
sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_KEEPIDLE, after_idle_sec)
sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_KEEPINTVL, interval_sec)
sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_KEEPCNT, max_fails)
```

**Review by mhthies — APPROVED** [review]
Thanks for reworking your code. That looks really good to me! :)

I'd like to test if we now need the timeout mechanism at all (see my new inline comment). Apart from that, you should take a look at the failing unittest and the MyPy complaints. Then, I'm fine with merging the PR.

File: knxdclient/client.py
Comment (mrctrifork) [reply knxdclient/client.py:160]: I tested the mechanism manually by connecting to a Debian docker image with KNXD installed and used `knxtool` to simulate some traffic which then I stopped and it seems to work.

It's possible that there is some stuff about coroutines / tasks that goes over my head as I am seeing the CI/CD failing.

The definitive test will come in 5h and 20 minutes :)

**Comment by mrctrifork** [comment]
[As I've stated in this thread](https://github.com/mhthies/knxdclient/pull/2#discussion_r1408397263) 

We'll have the official test some time from now :), once that happens I'll consider the new approval to be working.

And you're right. CI/CD is failing now so I'll take a look at that tomorrow

**Comment by mrctrifork** [comment]
I can confirm it did the trick!

File: knxdclient/client.py
Comment (mrctrifork) [reply knxdclient/client.py:160]: > Have you tested whether the event notification mechanism, you implemented today, is sufficient to detect the connection loss and stop the receive loop, when your KNXD server reboots – without the timeout here in the run() coroutine (or with a very very long timeout)?

It was enough as I kept receiving data throughout the night! 🎉

But I see how setting the TCP props on the socket will also be of use.

File: test/test_client.py
Comment (mhthies) [inline test/test_client.py]: ~~If a `ConnectionAbortedError` is raised here (in an explicit `stop()`, I would consider this a bug, so we should not ignore the exception.~~

EDIT: Sorry, I misread the diff. You already removed the suppressing here again. Then, everything is fine.

File: knxdclient/client.py
Comment (mhthies) [inline knxdclient/client.py]: Does this work when manually closing the connection via `close()` such that no exception is raised in this case?

**Comment by mhthies** [comment]
> I can confirm it did the trick!

Then, we can remove the timeout parameter, can't we?
We can also keep it, but than we should document in the docstring of the __init__method (via an `@param timeout ...` line) that it should not actually be required and *if* it is set, it will require regular KNX bus activity.

**Comment by mhthies** [comment]
I took the liberty to do some small improvements to the code style and add a test for the new behaviour. Please take a look, if this still is in your sense; then I will merge the PR.

**Comment by mrctrifork** [comment]
Good afternoon, I've been away from the computer for a bit. Thank you for your time to make these improvements and looking at the changes.

----

The other day I was answering your comment but I did not hit send in the end; shut down the machine and forget about everything.

> Then, we can remove the timeout parameter, can't we?
> We can also …

I meant that I just tried with my changes – didn't try with the TCP keepalive options

**Pull request merged**
